models/train.py

```diff
@@ -0,0 +1,236 @@
+                eval_end_slice = eval_begin_slice + max_steps
+                epoch = checkpoint['epoch']
+                writer.add_text('Checkpoint Event', f"Loaded checkpoint at {global_step} from '{checkpoint_filename}'. Training to {max_steps} steps.")
+        # End Checkpoint Loading
+
+        # Begin Data
+        writer.add_text('Data Event', "Initializing data loaders.")
+        train_filenames = sorted(glob.glob(train_path_str))[train_begin_slice:train_end_slice]
+        training_dataset = PackedDataset(train_filenames, n_chunks=chunks, block_size=block_size, shuffle=False)
+        training_dataloader = DataLoader(training_dataset, batch_size=batch_size, num_workers=num_workers, shuffle=False, pin_memory=True)
+
+        eval_filenames = sorted(glob.glob(eval_path_str))[eval_begin_slice:eval_end_slice]
+        eval_dataset = PackedDataset(eval_filenames, n_chunks=chunks, block_size=block_size, shuffle=False)
+        eval_dataloader = DataLoader(eval_dataset, batch_size=batch_size, num_workers=num_workers, shuffle=False, pin_memory=True)
+        writer.add_text('Data Event', "Initialized data loaders.")
+        # End Data
+
+        # Begin Train
+        writer.add_hparams(hparam_dict={
+                # Optimization
+                "lr":learning_rate,
+                "beta1":beta1,
+                "beta2":beta2,
+                "weight_decay":weight_decay,
+                "decay_learning_rate":decay_learning_rate,
+                "gradient_clip":gradient_clip,
+                
+                # Data loading
+                "training_dataset_path":train_path_str,
+                "eval_dataset_path":eval_path_str,
+                "num_workers":num_workers,
+                "batch_size":batch_size,
+                "micro_batch_size":micro_batch_size,
+                "block_size":block_size,
+                "chunk_size":chunks,
+
+                # Training
+                "epochs":epochs,
+                "warmup_steps":warmup_steps,
+                "max_steps":max_steps,
+                "train_begin_slice":train_begin_slice,
+                "train_end_slice":train_end_slice,
+                "eval_begin_slice":eval_begin_slice,
+                "eval_end_slice":eval_end_slice,
+            },metric_dict={},global_step=global_step)
+
+        model.train()
+        for epoch in range(epoch, epochs):
+            for step, td in enumerate(training_dataloader):
+                input_ids = td[:, :].contiguous().to(model.device)
+                labels = td[:, :].contiguous().to(model.device)
+
+                with profile(f"forward"):
+                    # HF implements target text shifting, or we can do it manually.
+                    m_output = model(input_ids, labels=labels) 
+                    loss = m_output.loss
+                with profile(f"backward"):
+                    loss.backward()
+                with profile(f"optimizer {type(optimizer).__name__}"):
```

Comment:
same comment as above, do all these profilers have overhead? We should maybe wrap them in a flag.